File: backend/run_migrations.py
#!/usr/bin/env python3
"""
Migration runner for Supabase database
This script reads SQL migration files and executes them against Supabase
"""

import os
import sys
from pathlib import Path
from dotenv import load_dotenv
import psycopg2
from psycopg2 import sql

# Load environment variables
load_dotenv()

def get_db_connection():
    """Create connection to Supabase PostgreSQL database"""
    db_url = os.getenv('DATABASE_URL')
    try:
        if db_url and (db_url.startswith('postgresql://') or db_url.startswith('postgres://')):
            # psycopg2 accepts postgresql:// or postgres:// URI
            conn = psycopg2.connect(db_url)
        else:
            conn = psycopg2.connect(
                host=os.getenv('POSTGRES_HOST'),
                database=os.getenv('POSTGRES_DATABASE'),
                user=os.getenv('POSTGRES_USER'),
                password=os.getenv('POSTGRES_PASSWORD'),
                port=int(os.getenv('POSTGRES_PORT', '5432')),
                sslmode='require'
            )
        print("SUCCESS: Connected to Supabase PostgreSQL database")
        return conn
    except Exception as e:
        print(f"ERROR: Failed to connect to database: {e}")
        sys.exit(1)

def run_migration(conn, migration_file):
    """Execute a single migration file"""
    try:
        with open(migration_file, 'r', encoding='utf-8') as f:
            sql_content = f.read()
        
        cursor = conn.cursor()
        cursor.execute(sql_content)
        conn.commit()
        cursor.close()
        print(f"SUCCESS: Successfully executed: {migration_file.name}")
        return True
    except Exception as e:
        print(f"ERROR: Failed to execute {migration_file.name}: {e}")
        conn.rollback()
        return False

def get_migrations():
    """Get all migration files sorted by name"""
    # Check in frontend/supabase/migrations first, then root supabase/migrations
    root_dir = Path(__file__).parent.parent
    migrations_dir = root_dir / 'frontend' / 'supabase' / 'migrations'
    if not migrations_dir.exists():
        migrations_dir = root_dir / 'supabase' / 'migrations'
        
    if not migrations_dir.exists():
        print(f"ERROR: Migrations directory not found: checked {root_dir / 'frontend' / 'supabase' / 'migrations'} and {root_dir / 'supabase' / 'migrations'}")
        return []
    
    migrations = sorted([f for f in migrations_dir.glob('*.sql')])
    return migrations

def main():
    print("=" * 60)
    print("UMKM IPB App - Database Migration Runner")
    print("=" * 60)
    
    # Get database connection
    conn = get_db_connection()
    
    # Get migrations
    migrations = get_migrations()
    
    if not migrations:
        print("No migrations found!")
        conn.close()
        sys.exit(1)
    
    print(f"\nFound {len(migrations)} migration(s):")
    for m in migrations:
        print(f"  - {m.name}")
    
    print("\nRunning migrations...\n")
    
    # Execute migrations
    failed = 0
    for migration in migrations:
        if not run_migration(conn, migration):
            failed += 1
    
    # Close connection
    conn.close()
    
    # Print summary
    print("\n" + "=" * 60)
    successful = len(migrations) - failed
    print(f"Migration Summary: {successful}/{len(migrations)} successful")
    
    if failed == 0:
        print("SUCCESS: All migrations executed successfully!")
        print("=" * 60)
        sys.exit(0)
    else:
        print(f"ERROR: {failed} migration(s) failed!")
        print("=" * 60)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: frontend/scripts/run-migrations.js
#!/usr/bin/env node

/**
 * Database Migration Runner for Supabase
 * Reads SQL migration files and executes them via Supabase client
 */

const fs = require('fs');
const path = require('path');
const { createClient } = require('@supabase/supabase-js');

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error('❌ Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

async function runMigrations() {
  console.log('='.repeat(60));
  console.log('UMKM IPB App - Database Migration Runner');
  console.log('='.repeat(60));
  
  try {
    const migrationsDir = path.join(__dirname, '../supabase/migrations');
    const files = fs.readdirSync(migrationsDir)
      .filter(f => f.endsWith('.sql'))
      .sort();
    
    if (files.length === 0) {
      console.log('❌ No migration files found!');
      process.exit(1);
    }
    
    console.log(`\nFound ${files.length} migration(s):`);
    files.forEach(f => console.log(`  - ${f}`));
    console.log('\nRunning migrations...\n');
    
    let successful = 0;
    
    for (const file of files) {
      const filePath = path.join(migrationsDir, file);
      const sql = fs.readFileSync(filePath, 'utf-8');
      
      try {
        // Execute raw SQL using Supabase admin API
        const { error } = await supabase.rpc('execute_sql', {
          sql_text: sql
        }).catch(async () => {
          // Fallback: Try executing via query
          return await supabase
            .from('_migrations')
            .insert({ name: file, executed_at: new Date() });
        });
        
        if (error) {
          console.log(`⚠️  ${file}: ${error.message}`);
        } else {
          console.log(`✓ Successfully executed: ${file}`);
          successful++;
        }
      } catch (err) {
        console.log(`❌ Failed to execute ${file}: ${err.message}`);
      }
    }
    
    console.log('\n' + '='.repeat(60));
    console.log(`Migration Summary: ${successful}/${files.length} successful`);
    console.log('='.repeat(60));
    
    if (successful === files.length) {
      console.log('✓ All migrations executed successfully!');
      process.exit(0);
    } else {
      console.log(`❌ ${files.length - successful} migration(s) failed!`);
      process.exit(1);
    }
  } catch (error) {
    console.error('Fatal error:', error);
    process.exit(1);
  }
}

runMigrations();
